diff --git a/Makefile b/Makefile
index 1c64dfd..8c2a533 100644
--- a/Makefile
+++ b/Makefile
@@ -40,7 +40,7 @@ clean_htslib:
 
 %.o: src/%.c
 	mkdir -p obj && \
-		$(CC) -c -pthread -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2 $(CFLAGS) \
+		$(CC) -c -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2 $(CFLAGS) \
 		-Isrc -Ihtslib $(EXTRA_CFLAGS) $^ -o $@
 
 .PHONY: clean_obj
@@ -49,10 +49,10 @@ clean_obj:
 
 
 modbam2bed: modbam2bed.o common.o counts.o bamiter.o args.o htslib/libhts.a
-	$(CC) -pthread -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2 $(CFLAGS) \
+	$(CC) -Wall -fstack-protector-strong -D_FORTIFY_SOURCE=2 $(CFLAGS) \
 		-Isrc -Ihtslib $(EXTRA_CFLAGS) $(EXTRA_LDFLAGS)\
 		$^ $(ARGP) \
-		-lm -lz -llzma -lbz2 -lpthread -lcurl -lcrypto $(EXTRA_LIBS) \
+		-lm -lz -lbz2 -lpthread $(EXTRA_LIBS) \
 		-o $(@)
 
 .PHONY: clean
