--- a/src/Makefile
+++ b/src/Makefile
@@ -106,16 +106,16 @@ incFiles = lastz.h infer_scores.h \
 
 
 lastz: $(foreach part,${srcFiles},${part}.o)
-	${CC} $(foreach part,${srcFiles},${part}.o) -lm -o $@
+	${CC} $(foreach part,${srcFiles},${part}.o) ${LDFLAGS} -lm -o $@.js
 
 lastz_D: $(foreach part,${srcFiles},${part}_D.o)
-	${CC} $(foreach part,${srcFiles},${part}_D.o) -lm -o $@
+	${CC} $(foreach part,${srcFiles},${part}_D.o) ${LDFLAGS} -lm -o $@.js
 
 lastz_32: $(foreach part,${srcFiles},${part}_32.o)
-	${CC} $(foreach part,${srcFiles},${part}_32.o) -lm -o $@
+	${CC} $(foreach part,${srcFiles},${part}_32.o) ${LDFLAGS} -lm -o $@.js
 
 lastz_40: $(foreach part,${srcFiles},${part}_40.o)
-	${CC} $(foreach part,${srcFiles},${part}_40.o) -lm -o $@
+	${CC} $(foreach part,${srcFiles},${part}_40.o) ${LDFLAGS} -lm -o $@.js
 
 # cleanup
 
