diff --git a/seqtk.c b/seqtk.c
index feaeda8..408c970 100644
--- a/seqtk.c
+++ b/seqtk.c
@@ -1598,6 +1598,7 @@ static int usage()
 
 int main(int argc, char *argv[])
 {
+	optind = 1;  // Reset optind so that we can support multiple calls to main()
 	if (argc == 1) return usage();
 	if (strcmp(argv[1], "comp") == 0) stk_comp(argc-1, argv+1);
 	else if (strcmp(argv[1], "fqchk") == 0) stk_fqchk(argc-1, argv+1);
