diff --git a/main.c b/main.c
index d8f7750d1..4933f6755 100644
--- a/main.c
+++ b/main.c
@@ -254,11 +254,6 @@ main(int argc, char **argv)
 	set_locale_stuff();
 
 	(void) signal(SIGSEGV, catchsig);
-	(void) signal(SIGFPE, catchsig);
-#ifdef SIGBUS
-	(void) signal(SIGBUS, catchsig);
-#endif
-
 	/*
 	 * Ignore SIGPIPE so that writes to pipes that fail don't
 	 * kill the process but instead return -1 and set errno.
