diff --git a/Makefile b/Makefile
index 160314a..f890d57 100644
--- a/Makefile
+++ b/Makefile
@@ -140,7 +140,7 @@ libbam.a:$(LOBJS)
 	$(AR) -csru $@ $(LOBJS)
 
 samtools: $(AOBJS) $(LZ4OBJS) libbam.a libst.a $(HTSLIB)
-	$(CC) $(ALL_LDFLAGS) -o $@ $(AOBJS) $(LZ4OBJS) libbam.a libst.a $(HTSLIB_LIB) $(CURSES_LIB) -lm $(ALL_LIBS) -lpthread
+	$(CC) $(ALL_LDFLAGS) -o ../build/$@.mjs $(AOBJS) $(LZ4OBJS) libbam.a libst.a $(HTSLIB_LIB) $(CURSES_LIB) -lm $(ALL_LIBS) -lpthread
 
 # For building samtools and its test suite only: NOT to be installed.
 libst.a: $(LIBST_OBJS)
diff --git a/bam_sort.c b/bam_sort.c
index 0bf346c..39c7837 100644
--- a/bam_sort.c
+++ b/bam_sort.c
@@ -2027,10 +2027,9 @@ static int sort_blocks(int n_files, size_t k, bam1_tag *buf, const char *prefix,
             w[i].no_save = 0;
         }
         pos += w[i].buf_len; rest -= w[i].buf_len;
-        pthread_create(&tid[i], &attr, worker, &w[i]);
     }
     for (i = 0; i < n_threads; ++i) {
-        pthread_join(tid[i], 0);
+        worker(&w[i]);
         if (w[i].error != 0) {
             errno = w[i].error;
             print_error_errno("sort", "failed to create temporary file \"%s.%.4d.bam\"", prefix, w[i].index);
diff --git a/version.sh b/version.sh
index 5ccd9bb..bb7032a 100755
--- a/version.sh
+++ b/version.sh
@@ -30,7 +30,7 @@ VERSION=1.10
 if [ -e .git ]
 then
     # If we ever get to 10.x this will need to be more liberal
-    VERSION=`git describe --match '[0-9].[0-9]*' --dirty --always`
+    VERSION=`git describe --match '[0-9].[0-9]*' --always`
 fi
 
 echo $VERSION
