diff --git a/Makefile b/Makefile
index c7acbf9..af46a84 100644
--- a/Makefile
+++ b/Makefile
@@ -132,7 +132,7 @@ version.h:
 	echo '#define SAMTOOLS_CC "$(CC)"' >> $@
 	echo '#define SAMTOOLS_CPPFLAGS "$(CPPFLAGS)"' >> $@
 	echo '#define SAMTOOLS_CFLAGS "$(CFLAGS)"' >> $@
-	echo '#define SAMTOOLS_LDFLAGS "$(LDFLAGS)"' >> $@
+	echo '#define SAMTOOLS_LDFLAGS "--preload-file examples/@/samtools/examples/ -s ERROR_ON_UNDEFINED_SYMBOLS=0 -O3"' >> $@
 	echo '#define SAMTOOLS_HTSDIR "$(HTSDIR)"' >> $@
 	echo '#define SAMTOOLS_LIBS "$(LIBS)"' >> $@
 	echo '#define SAMTOOLS_CURSES_LIB "$(CURSES_LIB)"' >> $@
@@ -150,7 +150,7 @@ LIBST_OBJS = sam_opts.o sam_utils.o bedidx.o bam.o
 
 
 samtools: $(AOBJS) $(LZ4OBJS) libst.a $(HTSLIB)
-	$(CC) $(ALL_LDFLAGS) -o $@ $(AOBJS) $(LZ4OBJS) libst.a $(HTSLIB_LIB) $(CURSES_LIB) -lm $(ALL_LIBS) -lpthread
+	$(CC) $(ALL_LDFLAGS) -o ../build/$@.mjs $(AOBJS) $(LZ4OBJS) libst.a $(HTSLIB_LIB) $(CURSES_LIB) -lm $(ALL_LIBS) -lpthread
 
 # For building samtools and its test suite only: NOT to be installed.
 libst.a: $(LIBST_OBJS)
diff --git a/bam_sort.c b/bam_sort.c
index 64e643f..148ddfe 100644
--- a/bam_sort.c
+++ b/bam_sort.c
@@ -3181,10 +3181,9 @@ static int sort_blocks(size_t k, bam1_tag *buf, const sam_hdr_t *h,
         in_mem[i].from = pos;
         in_mem[i].to = pos + w[i].buf_len;
         pos += w[i].buf_len; rest -= w[i].buf_len;
-        pthread_create(&tid[i], &attr, worker, &w[i]);
     }
     for (i = 0; i < n_threads; ++i) {
-        pthread_join(tid[i], 0);
+        worker(&w[i]);
         if (w[i].error != 0) {
             errno = w[i].error;
             print_error_errno("sort", "failed to sort block %d", i);
diff --git a/version.sh b/version.sh
index 10e4cb7..cb10c75 100755
--- a/version.sh
+++ b/version.sh
@@ -30,7 +30,7 @@ VERSION=1.24
 if [ -e .git -a "x`which git`" != "x" ]
 then
     # If we ever get to 10.x this will need to be more liberal
-    VERSION=`git describe --match '[0-9].[0-9]*' --dirty --always`
+    VERSION=`git describe --match '[0-9].[0-9]*' --always`
 fi
 
 echo $VERSION
